### PR TITLE
Lkt: use lkt_toolbox instead of lkt_resolve driver

### PR DESCRIPTION
**Pull request by thvnx**
For convenience, get rid of the lkt_resolve driver and use lkt_toolbox
to run Lkt's tests.

TN: UA29-022

### CODE REVIEW

File: contrib/lkt/extensions/mains/lkt_toolbox.adb [hunk 1]
```diff
@@ -87,9 +88,15 @@ begin
    if Arg.Parser.Parse then
       for File_Name of Arg.Files.Get loop
          declare
+            File_Name_Str : constant String := To_String (File_Name);
             Unit : constant Analysis.Analysis_Unit
```
Comment (pmderodat) [inline contrib/lkt/extensions/mains/lkt_toolbox.adb:92]: Please realign colons

File: langkit/templates/pkg_implementation_body_ada.mako [hunk 2]
```diff
@@ -3704,9 +3704,12 @@ package body ${ada_lib_name}.Implementation is
       if Ent.Node /= null then
          declare
             Node_Image : constant Text_Type := Short_Text_Image (Ent.Node);
+            File_Name : constant Text_Type := To_Text (
```
Comment (pmderodat) [inline langkit/templates/pkg_implementation_body_ada.mako:3707]: Please align the colon. Also please move the opening parent to the next line.

File: langkit/templates/pkg_implementation_body_ada.mako [hunk 3]
```diff
@@ -3704,9 +3704,12 @@ package body ${ada_lib_name}.Implementation is
...
+              Ada.Directories.Simple_Name (Get_Filename (Unit (Ent.Node))));
```
Comment (pmderodat) [inline langkit/templates/pkg_implementation_body_ada.mako:3708]: For clarity/brevity, I suggest using the dot notation as much as possible: `Ent.Node.Unit.Get_Filename`

File: langkit/templates/pkg_implementation_body_ada.mako [hunk 4]
```diff
@@ -3704,9 +3704,12 @@ package body ${ada_lib_name}.Implementation is
...
          begin
             return
-            (if Ent.Info.Rebindings /= null
+            (if Ent.Info.Rebindings /= null and File_Name /= "__prelude"
```
Comment (pmderodat) [inline langkit/templates/pkg_implementation_body_ada.mako:3711]: Should be `and then` (we almost never use `and` in Ada, except for bitwise operations)

File: testsuite/drivers/lkt_toolbox_driver.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import os
```
Comment (pmderodat) [inline testsuite/drivers/lkt_toolbox_driver.py:1]: Now that it is unused, could you also remove `lkt_resolve_driver.py`?

File: testsuite/drivers/lkt_toolbox_driver.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        return ['test.lkt']
+
+    @property
+    def script_and_args(self):
+        return ['lkt_toolbox', 'test.lkt']
+
+    def run(self):
+        derived_env = dict(os.environ)
+
+        if self.valgrind_enabled:
```
Comment (pmderodat) [inline testsuite/drivers/lkt_toolbox_driver.py:23]: I believe this is useless: `lkt_toolbox` does not check this environment variable, and we want `lkt_toolbox` to be run under Valgrind anyway in this case. See `run_and_check`'s `memcheck` argument. Also please pass `for_coverage=True`.

File: testsuite/drivers/lkt_toolbox_driver.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+
+class LktToolboxDriver(BaseDriver):
+    """
+    Custom driver that will automatically run the "lkt_toolbox" utility
+    on the testcase's "test.lkt" file.
+    """
+
+    @property
+    def mandatory_files(self):
```
Comment (pmderodat) [inline testsuite/drivers/lkt_toolbox_driver.py:13]: This property is unused: please remove it.

File: testsuite/drivers/lkt_toolbox_driver.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        return ['test.lkt']
+
+    @property
+    def script_and_args(self):
```
Comment (pmderodat) [inline testsuite/drivers/lkt_toolbox_driver.py:17]: `lkt_toolbox` is not a script, this is misleading. I would suggest to just get rid of this property and just put the list literal in `run`.

**Review by pmderodat — CHANGES_REQUESTED** [review]

File: langkit/templates/pkg_implementation_body_ada.mako (hunk 3 above)
Comment (thvnx) [reply langkit/templates/pkg_implementation_body_ada.mako]: Partially done, I couldn't use the dot notation for Get_Filename.

File: testsuite/drivers/lkt_toolbox_driver.py (hunk 5 above)
Comment (thvnx) [reply testsuite/drivers/lkt_toolbox_driver.py]: Done. I also removed `contrib/lkt/lkt_resolve.py`.

File: testsuite/drivers/lkt_toolbox_driver.py (hunk 6 above)
Comment (thvnx) [reply testsuite/drivers/lkt_toolbox_driver.py]: Thanks for the info. I removed the env variable and added memcheck/for_coverage=True.

File: langkit/templates/pkg_implementation_body_ada.mako (hunk 3 above)
Comment (pmderodat) [reply langkit/templates/pkg_implementation_body_ada.mako]: That’s unexpected: we do use dot notations for `Get_Filename` in the code base, and it should work everywhere… what happened when you tried?

Comment (thvnx) [reply langkit/templates/pkg_implementation_body_ada.mako]:
```
/langkit/contrib/python/build/src/libpythonlang-implementation.adb:6632:58: error: no selector "Get_Filename" for type "Analysis_Unit_Type" defined at libpythonlang-implementation.ads:10751
   [Ada]          liblktlang-generic_introspection.ads
/langkit/contrib/lkt/build/src/liblktlang-implementation.adb:9803:58: error: no selector "Get_Filename" for type "Analysis_Unit_Type" defined at liblktlang-implementation.ads:16768

   compilation of liblktlang-implementation.adb failed
```

Comment (pmderodat) [reply langkit/templates/pkg_implementation_body_ada.mako]: Oh!! I did not realize we were in Langkit Mako templates… This Lkt-specific code does not belong here: `__prelude` is a Lkt-specific thing, which does not necessarily exists in other Langkit-generated frontends.

IIRC, what we discussed live was to roll your own implementation of an `Image` in `lkt_toolbox.adb`

Comment (thvnx) [reply langkit/templates/pkg_implementation_body_ada.mako]: I remember, but I thought this was a possible solution, sorry about that, I'm going to fix it.

File: contrib/lkt/language/parser.py [hunk 9]
```diff
@@ -608,6 +608,17 @@ def assert_bare():
...
 
+    @langkit_property(public=True)
+    def as_bare_decl():
+        """
+        Get this declaration without rebindings information.
+        """
+        # This is used by lkt_toolbox as a simple solution to filter out
+        # rebindings information for the prelude declarations. TODO: improve it
+        # in order to properly show rebindings information for prelude
+        # declarations (i.e.: do not show sloc nor unit name).
```
Comment (pmderodat) [inline contrib/lkt/language/parser.py:619]: No need for `:` (`i.e. do not show sloc nor unit name`)

**Review by pmderodat — CHANGES_REQUESTED** [review]

**Review by pmderodat — APPROVED** [review]

**Pull request merged**